Hopefully/Procedure.cs

```diff
@@ -97,15 +97,22 @@ public static T Retry<T>(Func<T> what, out int failedAttempts, int attempts = 5)
                 {
                     return what();
                 }
-                catch
+                catch(Exception ex)
                 {
-                    attempts--;
-                    failedAttempts++;
-                    if (attempts == 0)
-                        throw;
+                    attempts = FailedAttempt(ref failedAttempts, attempts, wait, ex);   
                 }
             }
             return default(T); // Never reached
         }
+
+        private static int FailedAttempt(ref int failedAttempts, int attempts, TimeSpan wait, Exception ex)
+        {
+            attempts--;
+            failedAttempts++;
+            if (attempts == 0)
+                throw new ApplicationException("Reached the max number of attempts", ex);
```

Comment:
If we just `throw`, we keep the stack trace and you can have specialized catch statements based on the procedure being run. Why wrap it in an `ApplicationException`?